QWeb/internal/frame.py

```diff
@@ -207,7 +208,63 @@ def search_from_frames(driver=None, current_frame=None):
                 return web_element
             driver.switch_to.default_content()
             raise QWebTimeoutError('From frame decorator: Unable to locate element in given time')
-        return search_from_frames()
+
+        # pylint: disable=W0102
+        def search_from_frames_safari(driver=None, current_frame=None, parent_tree=[]):
+            keep_frame = kwargs.get('stay_in_current_frame', CONFIG['StayInCurrentFrame'])
+            if keep_frame:
+                return fn(*args, **kwargs)
+            err = None
+            if not driver:
+                driver = browser.get_current_browser()
+                driver.switch_to.default_content()
+            if current_frame is not None:
+                try:
+                    driver.switch_to.frame(current_frame)
+                    logger.debug('switching to childframe {}'.format(str(fn)))
+                except (StaleElementReferenceException, WebDriverException) as e:
+                    logger.warn(e)
+                    driver.switch_to.default_content()
+                    raise e
+            try:
+                web_element = fn(*args, **kwargs)
+            except QWebElementNotFoundError as e:
+                err = e
+                web_element = None
+            if is_valid(web_element):
+                return web_element
+            start = time.time()
+            timeout = CONFIG['FrameTimeout']
+            while time.time() < timeout + start:
```

Comment:
I don't understand the purpose of this while loop.
At the end of the loop: L261-263
Either an exception is raised or a web_element is returned which should mean that this while-statement acts like if-statement in reality

Reply:
Very old code, not much documentation. I wouldn't dare to touch it without exceptionally good reason and lot's of debugging and testing.

The basic idea is to catch and raise specific exceptions to timeout decorators. Certain exceptions make it to re-try finding elements from all frames etc. Timeout guarantees we will not get stuck, i.e jumps out of the loop once timeout expires.

There are also **no** version of keywords where it's ok that webelement is not found.